### ISSUE TITLE
[mlir] Lit tests fail after 2d2da25

### ISSUE THREAD

**Issue by phoebewang** [llvm/llvm-project#55992]
Hello, I just landed a patch to support FP16 emulation on X86: https://github.com/llvm/llvm-project/commit/2d2da259c8726fd5c974c01122a9689981a12196. Then I received a Buildbot notifaction from https://lab.llvm.org/buildbot#builders/61/builds/27616, where reported 2 failed tests:
```
Failed Tests (2):
  MLIR :: Integration/Dialect/SparseTensor/CPU/dense_output_f16.mlir
  MLIR :: Integration/Dialect/SparseTensor/CPU/sparse_sum_f16.mlir
```
I don't have any experiences on mlir and my local run show "Unsupported" on both tests.
Could anyone tell me how can I reproduce it locally, or help to generate a LLVM IR for me? Thanks!

**Comment by aartbik** [comment]
Does the patch change the ABI in any way?

We are trying to call

void *_mlir_ciface_addEltF16(void *, f16, ....

but perhaps the f16 has some "padding" applied, causing corruptions of all other parameters.
We observed a similar problem with complex on two f32. So rather than having one

    void *_mlir_ciface_addEltC32(void *coo, c, complex32 c32, ...

we "fixed" this as

    void *_mlir_ciface_addEltC32(void *coo, float r, float i,
                                 StridedMemRefType<index_type, 1> *iref,
                                 StridedMemRefType<index_type, 1> *pref) {
  return _mlir_ciface_addEltC32ABI(coo, complex32(r, i), iref, pref);

See around

https://github.com/llvm/llvm-project/blob/e16d13322b2617843511aebb29a502166824b07a/mlir/lib/ExecutionEngine/SparseTensorUtils.cpp#L1746

**Comment by phoebewang** [comment]
Thanks @joker-eph @aartbik 
> Does the patch change the ABI in any way?

If you are asking the `half` type from LLVM IR level, yes, it does. See the difference in the test https://github.com/llvm/llvm-project/commit/2d2da259c8726fd5c974c01122a9689981a12196#diff-b377175c48a86b089bb448538c7849573d52ede1488dfd067f48da683c0110e7L16
Before this patch, `half` was passed by GPR registers, after it, it will be passed by FPR(XMM) registers.

> but perhaps the f16 has some "padding" applied, causing corruptions of all other parameters.

Not padding but register class. IIUC, the problem is the jitted code passes the data in FPR while the ExecutionEngine expected the data is in GPR? If this is true, I think it's a problem of ExecutionEngine that has false ABI assumption on half type: https://github.com/llvm/llvm-project/blob/main/mlir/lib/ExecutionEngine/Float16bits.cpp#L33

Previously, Clang doesn't have ABI for half type and blocks arguments passing for it. The implementation above looks like a workaround to me. My patch is to enable the ABI and (will) allow pass half type by `_Float16`. So I think the ExecutionEngine may need improve to detect if `_Float16` is supported from cmake and use `_Float16` when possible.

In conclusion, I think I can land the patch and disable both tests temporaryly. I'll put comments to point to this issue, but I may not be able to fix it. WDYT?

**Comment by joker-eph** [comment]
It's not clear to me that we need to think in terms of register class here. I see it as an LLVM IR problem: we write our runtime in C/C++ and we are trying to interact with it when generating LLVM IR ourselves from a non-C source (MLIR).

The question to me is whether we can generate the right signatures for our runtime in LLVM IR that matches what clang will generate when compiling the runtime.
If the mapping is platform dependent, we better go through memory (like we do with memref descriptor).

**Comment by aartbik** [comment]
I am okay if you disable the two tests for now just so you can make progress. Please add this bug somewhere in the comment when you do so, so we can find a proper fix afterwards.

**Comment by aartbik** [comment]
Note that https://reviews.llvm.org/D128018 fixes the ABI issues hopefully once and for all (since we use memref's now for all non-trivially typed data). We should be able to re-enable the tests again.
